### DISCUSSION TITLE
Dynamic/relative asset paths

### DISCUSSION THREAD

**Discussion by ghost** [vikejs/vike#603] [Discussion] [answered]
Good day!

Hope you guys are all doing well. 

I have a question in regards to dynamic/relative asset paths. We are using vite-plugin-ssr in combination with an express server, meaning we are using SSR. Now, our software is built once and distributed to various environments. Some of which are changing the express router path and mount everything underneath a different path (instead of root) like `/test/` or `/doc/`.

During the build process we also run `vite build` and here we run into an issue. We simply don't know where the package will end up and are therefore not able to provide a proper `base` path. Now, I read the official vite documentation and they actually do support a relative base setting (or no setting at all) like `./`. Unfortunately this will lead to an assertion in vite-plugin-ssr.

Next I tried to use the experimental `renderBuiltUrl`, which works for css/svg etc but not for the js chunks etc.

After that I figured I check out the `server` directory and found a `vite-plugin-ssr.json` file which contains an object with various settings , after a build it looks like this: 

```
{
  "version": "0.4.39",
  "usesClientRouter": false,
  "baseUrl": "/",
  "baseAssets": null,
  "includeAssetsImportedByServer": false
}
```
changing the `baseAssets` to this 

```
{
  "version": "0.4.39",
  "usesClientRouter": false,
  "baseUrl": "/",
  "baseAssets": "/documentation",
  "includeAssetsImportedByServer": false
}
```

results in a stack trace (I am aware that my change is probably not something that was expected :D).

```
Error: [vite-plugin-ssr@0.4.39][Bug] You stumbled upon a bug in vite-plugin-ssr's source code. Reach out at https://github.com/brillout/vite-plugin-ssr/issues/new or https://discord.com/invite/qTq92FQzKb and include this error stack (the error stack is usually enough to fix the problem). A maintainer will fix the bug (usually under 24 hours). Do not hesitate to reach out as it makes vite-plugin-ssr more robust.
    at assertPluginManifest (C:\Users\ChristophLipp\Desktop\mdd\node_modules\vite-plugin-ssr\dist\cjs\node\plugin\plugins\manifest\assertPluginManifest.js:13:24)
    at getGlobalContext (C:\Users\ChristophLipp\Desktop\mdd\node_modules\vite-plugin-ssr\dist\cjs\node\globalContext.js:33:57)
    at async initializePageContext (C:\Users\ChristophLipp\Desktop\mdd\node_modules\vite-plugin-ssr\dist\cjs\node\renderPage.js:125:27)
    at async renderPage_ (C:\Users\ChristophLipp\Desktop\mdd\node_modules\vite-plugin-ssr\dist\cjs\node\renderPage.js:30:41)
    at async renderPage (C:\Users\ChristophLipp\Desktop\mdd\node_modules\vite-plugin-ssr\dist\cjs\node\renderPage.js:155:16)
    at async C:\Users\ChristophLipp\Desktop\mdd\server\api\modules\rendering_pages_v2.js:181:29
```

Is there anything I forgot to try or is there something that we can configure that might help with our setup? Thanks very much and have a great day!

Cheers,
Christoph

**Comment by brillout** [comment, marked as answer]
> I read the official vite documentation and they actually do support a relative base setting [...] like `./`

How exactly would that help and how would you use it? Can you give an example of assets for which you want to have a relative base and why?

A challenge with relative paths is that they depend on the current URL. That's why VPS doesn't support them for now. (I'm not sure how/if VPS can support them.)

**Reply by ghost** [reply]
The idea is the following:

As of now, every asset path after a build points to `/assets/.....`.  Now, running the application under `/` works as expected. But, if we change the mount path of the express router to `/docs` the asset path in the HTML should be `/docs/assets......`. 

I.E:
```
const app = express();
const router = new express.Router();
// here we configure everything on the router, including vite-plugin-ssr
app.use("/docs", router); // <-- This changes based on the environment the software is running on
```

But, if all the assets would point to `./assets/...` the browser would load every asset based off the current URL.

Samples:
- Browser hits `http://localhost`; Browser would fetch `/assets`
- Browser hits `/http://localhost/docs/`; Browser would fetch `/docs/assets`

I believe that's the very reason vite itself does allow setting `base: './'` in `vite.config.ts`, so everything can be pushed to different hosts/base paths. Hope this makes sense :)

**Reply by brillout** [reply]
Vite transpiles this:

```js
import logo from './logo.svg'
```

to:

```js
const logo = '/assets/logo.svg'
````

If you want Vite to transpile it to this instead:

```js
const logo = './assets/logo.svg'
````

Then this works only for `https://my-website.com/docs/` but not for `https://my-website.com/docs/some-nested-path`, because:

- `<img src="./assets/logo.svg" />` at `https://my-website.com/docs/` is resolved to `https://my-website.com/docs/assets/logo.svg` => `200`
- `<img src="./assets/logo.svg" />` at `https://my-website.com/docs/some-nested-path` is resolved to `https://my-website.com/docs/some-nested-path/assets/logo.svg` => `404`

I'm not sure how to solve this.

Correct me if I'm wrong, but I think Vite's support for base relative paths only works for SPAs that are served under a single URL.

Feel free to play around with vanilla Vite (without VPS) in [MPA mode](https://vitejs.dev/config/shared-options.html#apptype) and see whether it works (I never tried but I suspect it doesn't.)

**Reply by ghost** [reply]
Fair enough! That is a very good question and I myself have not checked how it is being solved (`.../assets` maybe?). 
Either way, everything you said makes sense to me and makes me think changing our deployment strategy might be the only option.

Thanks for your thoughts and opinion on this.

**Reply by brillout** [reply]
Actually, it may be possible using [`<base>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base).

I created a feature request for relative base paths: https://github.com/brillout/vite-plugin-ssr/issues/606.

**Reply by brillout** [reply]
Btw., if your company is up for it: https://github.com/sponsors/brillout.

**Reply by ghost** [reply]
Thanks I will take this to the higher ups :) In the meantime I have sent you a small amount as a token of appreciation from my personal account @ser1us. Thanks again and have a drink on me!

**Reply by brillout** [reply]
Thanks 💚.

> have a drink on me

I will, with a bavarian beer :-).
